### PR TITLE
Start of updating CAN ID header format to match v1.0

### PR DESCRIPTION
**Pull request by aasmune**
This is a start of updating the ID header to match the new format for v1.0. Note: It could be that I have missed some parts which also must be updated related to the new header format, or that I have made some mistakes. Therefore, I would appreciate that you check my changes thoroughly. 

These changes were originally included in #96, but we chose to split it out to a separate PR.

### CODE REVIEW

File: canard.c [hunk 1]
```diff
@@ -37,15 +37,14 @@
...
 
 #define TRANSFER_ID_BIT_LEN                         5U
-#define ANON_MSG_DATA_TYPE_ID_BIT_LEN               2U
 
-#define SOURCE_ID_FROM_ID(x)                        ((uint8_t) (((x) >> 0U)  & 0x7FU))
-#define SERVICE_NOT_MSG_FROM_ID(x)                  ((bool)    (((x) >> 7U)  & 0x1U))
-#define REQUEST_NOT_RESPONSE_FROM_ID(x)             ((bool)    (((x) >> 15U) & 0x1U))
+#define SOURCE_ID_FROM_ID(x)                        ((uint8_t) (((x) >> 1U)  & 0x7FU))
+#define SERVICE_NOT_MSG_FROM_ID(x)                  ((bool)    (((x) >> 25U) & 0x1U))
+#define REQUEST_NOT_RESPONSE_FROM_ID(x)             ((bool)    (((x) >> 26U) & 0x1U))
```
Comment (pavel-kirienko) [inline canard.c]: the correct offset is 24, not 26

File: canard.c [hunk 2]
```diff
@@ -37,15 +37,14 @@
...
 #define DEST_ID_FROM_ID(x)                          ((uint8_t) (((x) >> 8U)  & 0x7FU))
-#define PRIORITY_FROM_ID(x)                         ((uint8_t) (((x) >> 24U) & 0x1FU))
+#define PRIORITY_FROM_ID(x)                         ((uint8_t) (((x) >> 26U) & 0x1FU))
 #define MSG_TYPE_FROM_ID(x)                         ((uint16_t)(((x) >> 8U)  & 0xFFFFU))
```
Comment (pavel-kirienko) [inline canard.c]: We don't have message ID anymore; this macro should be renamed into "subject ID".

The mask should be 0x7FFF, because the subject ID ranges from 0 to 32767 inclusive. This [was discussed on the forum](https://forum.uavcan.org/t/alternative-transport-protocols/324) but the spec draft was kept outdated for a while (until now; I just pushed this fix to the master of the spec repo).

File: canard.c [hunk 3]
```diff
@@ -37,15 +37,14 @@
...
-#define SRV_TYPE_FROM_ID(x)                         ((uint8_t) (((x) >> 16U) & 0xFFU))
+#define SRV_TYPE_FROM_ID(x)                         ((uint8_t) (((x) >> 15U) & 0x1FFU))
 
 #define MAKE_TRANSFER_DESCRIPTOR(data_type_id, transfer_type, src_node_id, dst_node_id)             \
     (((uint32_t)(data_type_id)) | (((uint32_t)(transfer_type)) << 16U) |                            \
```
Comment (pavel-kirienko) [inline canard.c]: We don't have data type ID anymore. We have *port ID*, which in turn can be either *service ID* or *subject ID*.

File: canard.c [hunk 4]
```diff
@@ -151,29 +150,19 @@ int16_t canardBroadcast(CanardInstance* ins,
     uint32_t can_id = 0;
     uint16_t crc = 0xFFFFU;
 
+    can_id = ((uint32_t) priority << 26U) | ((uint32_t) data_type_id << 8U) | (uint32_t) canardGetLocalNodeID(ins);
```
Comment (pavel-kirienko) [inline canard.c]: `((uint32_t) canardGetLocalNodeID(ins) << 1)`

File: canard.c [hunk 5]
```diff
@@ -151,29 +150,19 @@ int16_t canardBroadcast(CanardInstance* ins,
...
+
     if (canardGetLocalNodeID(ins) == 0)
```
Comment (pavel-kirienko) [inline canard.c]: Node ID of zero is no longer a special case.

File: canard.c [hunk 6]
```diff
@@ -211,9 +200,9 @@ int16_t canardRequestOrRespond(CanardInstance* ins,
         return -CANARD_ERROR_NODE_ID_NOT_SET;
     }
 
-    const uint32_t can_id = ((uint32_t) priority << 24U) | ((uint32_t) data_type_id << 16U) |
-                            ((uint32_t) kind << 15U) | ((uint32_t) destination_node_id << 8U) |
-                            (1U << 7U) | (uint32_t) canardGetLocalNodeID(ins);
+    const uint32_t can_id = ((uint32_t) priority << 26U) | ((uint32_t) data_type_id << 15U) |
+                            ((uint32_t) kind << 24U) | ((uint32_t) destination_node_id << 8U) |
+                            (1U << 25U) | (uint32_t) canardGetLocalNodeID(ins);
```
Comment (pavel-kirienko) [inline canard.c]: Source node ID not shifted.

File: tests/test_rxerr.cpp [hunk 7]
```diff
@@ -26,35 +26,34 @@
...
-#define CONSTRUCT_SNM_BIT (1 << 7)
-
-#define CONSTRUCT_ANON_MSG_ID(prio, disc, mtid, sid)    (CONSTRUCT_PRIO(prio) | \
-                                                         CONSTRUCT_DISC_MTID(mtid) | \
-                                                         CONSTRUCT_DISC(disc) | \
-                                                         CONSTRUCT_SID(sid) \
+#define CONSTRUCT_PRIO(prio)                    (((prio) & 0x7) << 26)
+#define CONSTRUCT_SUBJECT_ID(subject_id)        (((subject_id) & 0xFFFF) << 8)
+#define CONSTRUCT_SOURCE_ID(source_id)          (((source_id) & 0x7F) << 1)
+#define CONSTRUCT_SERVICE_ID(service_tid)      (((service_tid) & 0x1FF) << 15)
```
Comment (pavel-kirienko) [inline tests/test_rxerr.cpp]: What does TID mean?

File: canard.c (hunk 5 above)
Comment (aasmune) [reply canard.c]: What should be done instead? Should only bit 24 be set, remove the inner if (with error) and move the inner if in else out of the if/else statement?

File: tests/test_rxerr.cpp (hunk 7 above)
Comment (aasmune) [reply tests/test_rxerr.cpp]: Only a leftover from when I renamed the parameters, will be renamed to `service_id`

File: canard.c (hunk 5 above)
Comment (pavel-kirienko) [reply canard.c]: We just need to reserve some other node ID value at the implementation level. For example, we could treat every node ID above 127 as "anonymous". The discriminator is now also gone, instead we have "pseudo-ID" now (which is semantically similar to discriminator), see 4.4.1 https://new.uavcan.org/specification/UAVCAN_Specification_v1.0.pdf

Comment (aasmune) [reply canard.c]: I must have missed that section earlier... Will add a check which treats every node ID above 127 as anonymous as you proposed along with the pseudo-ID.

Comment (aasmune) [reply canard.c]: Not directly the scope of this PR, but during my work of updating the error handling related to anonymous transfers, I came over the error code CANARD_ERROR_NODE_ID_NOT_SET. This is currently only used if the user tries to transmit a multi-frame message or service request/response while in anonymous mode. Previously, this error was descriptive as "missing" node ID value was the cause. However, with the new specification this is no longer the truth as a node can have set an ID, but still be in "anonymous mode".

In my opinion, it should, therefore, be renamed to a more descriptive name e.g. CANARD_ERROR_INVALID_OPERATION_ANONYMOUS_NODE or something similar. Any comments to this?

Comment (pavel-kirienko) [reply canard.c]: > Previously, this error was descriptive as "missing" node ID value was the cause. However, with the new specification this is no longer the truth as a node can have set an ID, but still be in "anonymous mode".

This semantic was not supposed to be changed. Could you please elaborate how a node can be in the passive (anonymous) mode while having a valid node ID?

I should also note that the default node ID value should be changed: it used to be zero (anonymous), now it should be any value above 127.

Comment (aasmune) [reply canard.c]: What I meant is that previously the default node ID was 0 and this meant that it would transmit in anonymous mode unless the nodeID was updated with a non-zero value. With my previous comment, I forgot that we also need to change the default node ID from zero to a value above 127. Hence, a node could transmit in "normal" (non-anonymous) mode even though the node ID wasn't set by the user. However, when the default value is changed to a value above 127 (I propose 255), this is no longer the case, and my previous comment becomes "invalid".

Comment (aasmune) [reply canard.c]: Should we treat all node IDs above 127 as anonymous, or restrict it to a single value? After looking at it, I would propose to initialize the node ID to a value above 127 (let's say 255) and then only allow the user to update the value if it is between 0 and 127. This will be similar to how it is done today in canardSetLocalNodeID. In case an invalid value is provided, we can then return an error code.

Comment (pavel-kirienko) [reply canard.c]: Agree.

Comment (aasmune) [reply canard.c]: The above changes have now been implemented, ready for a new review.

File: canard.c [hunk 8]
```diff
@@ -110,20 +109,21 @@ void* canardGetUserReference(CanardInstance* ins)
     return ins->user_reference;
 }
 
-void canardSetLocalNodeID(CanardInstance* ins, uint8_t self_node_id)
+int16_t canardSetLocalNodeID(CanardInstance* ins, uint8_t self_node_id)
 {
     CANARD_ASSERT(ins != NULL);
 
     if ((ins->node_id == CANARD_BROADCAST_NODE_ID) &&
-        (self_node_id >= CANARD_MIN_NODE_ID) &&
```
Comment (aasmune) [inline canard.c]: Removed due to warning during compiling tests:
```bash
/home/aasmune/Documents/git/aasmune/libcanard/canard.c: In function ‘canardSetLocalNodeID’:
/home/aasmune/Documents/git/aasmune/libcanard/canard.c:117:23: error: comparison is always true due to limited range of data type [-Werror=type-limits]
         (self_node_id >= CANARD_MIN_NODE_ID) &&
                       ^~
cc1: all warnings being treated as errors
```

File: canard.c [hunk 9]
```diff
@@ -151,29 +151,20 @@ int16_t canardBroadcast(CanardInstance* ins,
...
-        {
-            return -CANARD_ERROR_INVALID_ARGUMENT;
-        }
-
-        // anonymous transfer, random discriminator
-        const uint16_t discriminator = (uint16_t)((crcAdd(0xFFFFU, payload, payload_len)) & 0x7FFEU);
-        can_id = ((uint32_t) priority << 24U) | ((uint32_t) discriminator << 9U) |
-                 ((uint32_t) (data_type_id & DTIDMask) << 8U) | (uint32_t) canardGetLocalNodeID(ins);
+        uint8_t pseudo_id = crcAdd(0xFFFFU, payload, payload_len) & 0x7FU;
+        can_id |= (1U << 24U) | ((uint32_t) pseudo_id << 1U);
```
Comment (pavel-kirienko) [inline canard.c]: `pseudo_id` does not affect the final value of `can_id`. Look at the branching logic, when we end up here, the bits 8...1 are already set to 0xFF.

File: canard.c [hunk 10]
```diff
@@ -151,29 +151,20 @@ int16_t canardBroadcast(CanardInstance* ins,
     uint32_t can_id = 0;
     uint16_t crc = 0xFFFFU;
 
-    if (canardGetLocalNodeID(ins) == 0)
+    can_id = ((uint32_t) priority << 26U) | ((uint32_t) data_type_id << 8U) | ((uint32_t) canardGetLocalNodeID(ins) << 1U);
```
Comment (pavel-kirienko) [inline canard.c]: If the local node ID is 0xFF, the shift operation overwrites the least significant bit of the subject ID.

File: canard.c [hunk 11]
```diff
@@ -206,14 +197,15 @@ int16_t canardRequestOrRespond(CanardInstance* ins,
     {
         return -CANARD_ERROR_INVALID_ARGUMENT;
     }
-    if (canardGetLocalNodeID(ins) == 0)
+    if (canardGetLocalNodeID(ins) == CANARD_BROADCAST_NODE_ID) 
     {
+        // Service transfer not supported for anonymous nodes 
         return -CANARD_ERROR_NODE_ID_NOT_SET;
     }
```
Comment (pavel-kirienko) [inline canard.c]: There is a problem with the original code: we don't seem to be checking `destination_node_id`, `inout_transfer_id`, `priority`, `kind`, and `payload_len` (against `CANARD_MAX_TRANSFER_PAYLOAD_LEN`). This also applies to other API functions, perhaps we should be more stringent? Or it could be implemented in a dedicated PR?

File: .gitignore [hunk 12]
```diff
@@ -34,6 +34,12 @@
 # CMake build directory
 build/
 cmake-build-*/
+CMakeFiles/
+cmake_install.cmake
+CMakeCache.txt
+Makefile
+demo
+run_tests
```
Comment (pavel-kirienko) [inline .gitignore]: This seems a bit drastic. I would recommend that you move your build files into a separate directory, e.g. `build`, which is already gitignored.

File: canard.c (hunk 11 above)
Comment (aasmune) [reply canard.c]: I would suggest move it to another PR as this one is mainly focused on updating the header format

File: .gitignore (hunk 12 above)
Comment (aasmune) [reply .gitignore]: All these files are created by running the travis commands for GCC 7 locally for Libcanard.
```bash
CC=gcc-7 && CXX=g++-7 && cd tests/ && cmake . && make
./run_tests --rng-seed time
```

In my opinion, these files should therefore be included in the gitignore as they will be present for all peoples which want to run tests locally before committing code.

Comment (pavel-kirienko) [reply .gitignore]: I didn't expect anyone to run them locally. The Travis script should be updated then.

**Comment by aasmune** [comment]
> The function "canardCleanupStaleTransfers()" must be removed (this affects the API and the demo); see section 4.2.1 for details. This change to the spec was implemented in order to improve deterministic behavior at the cost of somewhat increased average-case memory usage (the worst case is unaffected). The constant `CANARD_RECOMMENDED_STALE_TRANSFER_CLEANUP_INTERVAL_USEC` is therefore not needed either (there could be other constants which we no longer need).
> 
> Data type signature must be removed from everywhere. This includes the callback signature `CanardShouldAcceptTransfer`.
> 
> The API functions still use "data type ID" instead of "subject ID" and "service ID", this should be fixed.
> 
> The transfer priority constants require updating. We have proper mnemonic names for the eight priority levels defined in 4.1.4.
> 
> The script `show_data_type_info.py` will no longer be needed, since there are no signatures to worry about.

I would prefer to do this in a separate PR, for the same reason as the other comment, as this one is mainly for updating header format. I would suggest doing it as a part of #98.

**Comment by aasmune** [comment]
I notice that the travis build has been queued for a long time, could you check that I didn't make any mistakes when updating the travis file? @pavel-kirienko

**Comment by pavel-kirienko** [comment]
> I notice that the travis build has been queued for a long time, could you check that I didn't make any mistakes when updating the travis file?

It's weird. The changes seem benign and I can't imagine why would they cause any problems. I just tried restarting the build but it didn't help, it keeps getting stuck at the stage "Queued". Could be related to this:

- https://github.com/travis-ci/travis-ci/issues/9088
- https://www.traviscistatus.com/incidents/my5wm56npf7q

**Comment by aasmune** [comment]
> > I notice that the travis build has been queued for a long time, could you check that I didn't make any mistakes when updating the travis file?
> 
> It's weird. The changes seem benign and I can't imagine why would they cause any problems. I just tried restarting the build but it didn't help, it keeps getting stuck at the stage "Queued". Could be related to this:
> 
> * [travis-ci/travis-ci#9088](https://github.com/travis-ci/travis-ci/issues/9088)
> * https://www.traviscistatus.com/incidents/my5wm56npf7q

It could be. The build is finally finished queueing and has passed now!

**Comment by aasmune** [comment]
What do you think, are we ready to merge soon?

**Comment by pavel-kirienko** [comment]
Yes. Is this ready for another review?

**Comment by aasmune** [comment]
Yes, it is.

**Comment by aasmune** [comment]
> LGTM, assuming that the CRC handling and #98 will be updated separately.

Yes.
